### PR TITLE
[core] Add longer retry for dashboard agent on connecting with GCS

### LINKED ISSUE #41675: [Core] OSX gcs fault tolerance tests are flaky
### What happened + What you expected to happen

Screenshot 2023-12-07 at 7 51 46 AM

### Versions / Dependencies

master

### Reproduction script

n/a

### Issue Severity

None

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
When dashboard agent taking a long time to connect to GCS (default, > 5 secs), it will exit, which would also brings down the raylet. 
With GCS FT, it's not desirable to have dashboard agent terminates the raylet like this. This shows up in a test where detached actor was killed unexpected because the corresponding raylet dies as a result of dashboard agent failing to connect to GCS at start up. 

This PR makes dashboard agent retry more (30 secs instead of 5) by default when GCS client fails to connect to GCS. 

Closes https://github.com/ray-project/ray/issues/41675

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rickyyx** [comment]
Kicking off mac tests as well. 

A prototype test run passes on the draft PR: https://buildkite.com/ray-project/oss-ci-build-pr/builds/44311#018c605b-2807-4b21-89e6-1c9936acf511

**Review by rkooo567 — APPROVED** [review]
Nice investigation :)! Guess we don't need to cherry pick?

**Comment by rickyyx** [comment]
> Nice investigation :)! Guess we don't need to cherry pick?

yeah, maybe not a regression.

**Comment by rickyyx** [comment]
Somehow I am not seeing the medium A-J job started - cc @can-anyscale is it because of some optional CI flag?

File: dashboard/consts.py [hunk 1]
```diff
@@ -9,6 +9,10 @@
 DASHBOARD_AGENT_CHECK_PARENT_INTERVAL_S = env_integer(
     DASHBOARD_AGENT_CHECK_PARENT_INTERVAL_S_ENV_NAME, 0.4
 )
+DASHBOARD_AGENT_NUM_GCS_RECONNECT_RETRY = env_integer(
+    "DASHBOARD_AGENT_NUM_GCS_RECONNECT_RETRY",
```
Comment (iycheng) [inline dashboard/consts.py:13]: should we change it to 
https://github.com/ray-project/ray/blob/master/src/ray/common/ray_config_def.h#L65C7-L65C7 ?

**Review by fishbone — APPROVED** [review]
LG

**Comment by rickyyx** [comment]
Closes https://github.com/ray-project/ray/pull/41846

**Pull request merged**